R-package/src/cmake/modules/FindLibR.cmake

```
@@ -0,0 +1,229 @@
+#
+# Defines the following:
+#  LIBR_FOUND
+#  LIBR_HOME
+#  LIBR_EXECUTABLE
+#  LIBR_INCLUDE_DIRS
+#  LIBR_LIB_DIR
+#  LIBR_CORE_LIBRARY
+# and a cmake function to create R.lib for MSVC
+
+if(NOT R_ARCH)
+  if("${CMAKE_SIZEOF_VOID_P}" STREQUAL "4")
+    set(R_ARCH "i386")
+  else()
+    set(R_ARCH "x64")
+  endif()
+endif()
+
+if(NOT ("${R_ARCH}" STREQUAL "x64"))
+  message(FATAL_ERROR "LightGBM's R package currently only supports 64-bit operating systems")
+endif()
+
+# Creates R.lib and R.def in the build directory for linking with MSVC
+function(create_rlib_for_msvc)
+
+  # various checks and warnings
+  if(NOT WIN32 OR NOT MSVC)
+    message(FATAL_ERROR "create_rlib_for_msvc() can only be used with MSVC")
+  endif()
+
+  if(NOT EXISTS "${LIBR_LIB_DIR}")
+    message(FATAL_ERROR "LIBR_LIB_DIR was not set!")
+  endif()
+
+  find_program(GENDEF_EXE gendef)
+  find_program(DLLTOOL_EXE dlltool)
+
+  if(NOT GENDEF_EXE OR NOT DLLTOOL_EXE)
+    message(FATAL_ERROR "\nEither gendef.exe or dlltool.exe not found!\
+      \nDo you have Rtools installed with its MinGW's bin/ in PATH?")
+  endif()
+
+  # extract symbols from R.dll into R.def and R.lib import library
+  execute_process(COMMAND gendef
+    "-" "${LIBR_LIB_DIR}/R.dll"
+    OUTPUT_FILE "${CMAKE_CURRENT_BINARY_DIR}/R.def")
+  execute_process(COMMAND dlltool
+    "--input-def" "${CMAKE_CURRENT_BINARY_DIR}/R.def"
+    "--output-lib" "${CMAKE_CURRENT_BINARY_DIR}/R.lib")
+endfunction(create_rlib_for_msvc)
+
+# R version information is used to search for R's libraries in
+# the registry on Windows. Since this code is orchestrated by
+# an R script (src/install.libs.R), that script uses R's built-ins to
+# find the version of R and pass it through as a cmake variable
+if(CMAKE_R_VERSION)
+  message("R version passed into FindLibR.cmake: ${CMAKE_R_VERSION}")
+else()
+  message(FATAL_ERROR "Expected CMAKE_R_VERSION to be passed in but none was provided. Check src/install.libs.R")
+endif()
+
+# detection for OSX
+if(APPLE)
+
+  find_library(LIBR_LIBRARIES R)
+
+  if(LIBR_LIBRARIES MATCHES ".*\\.framework")
+    set(LIBR_HOME "${LIBR_LIBRARIES}/Resources" CACHE PATH "R home directory")
+    set(LIBR_INCLUDE_DIRS "${LIBR_HOME}/include" CACHE PATH "R include directory")
+    set(LIBR_EXECUTABLE "${LIBR_HOME}/R" CACHE PATH "R executable")
+    set(LIBR_LIB_DIR "${LIBR_HOME}/lib" CACHE PATH "R lib directory")
+  else()
+    get_filename_component(_LIBR_LIBRARIES "${LIBR_LIBRARIES}" REALPATH)
+    get_filename_component(_LIBR_LIBRARIES_DIR "${_LIBR_LIBRARIES}" DIRECTORY)
+    set(LIBR_EXECUTABLE "${_LIBR_LIBRARIES_DIR}/../bin/R")
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--vanilla" "-e" "cat(R.home())"
+      OUTPUT_VARIABLE LIBR_HOME
+    )
+    set(LIBR_HOME ${LIBR_HOME} CACHE PATH "R home directory")
+    set(LIBR_INCLUDE_DIRS "${LIBR_HOME}/include" CACHE PATH "R include directory")
+    set(LIBR_LIB_DIR "${LIBR_HOME}/lib" CACHE PATH "R lib directory")
```

To deduplicate code.
https://github.com/jameslamb/LightGBM/blob/2858c1c4a0e1c91f564c0222b815e42f2dd1c0ec/R-package/src/cmake/modules/FindLibR.cmake#L210-L214
```suggestion
    set(LIBR_INCLUDE_DIRS "${LIBR_HOME}/include")
    set(LIBR_LIB_DIR "${LIBR_HOME}/lib")
```

---

I'm not sure about this one @StrikerRUS . I think we still need to set  `LIBR_HOME`, since it is used further down.

Did you mean to just remove the `CACHE PATH` stuff since that's done all at the bottom? https://github.com/jameslamb/LightGBM/blob/feat/r-header/R-package/src/cmake/modules/FindLibR.cmake#L210

---

> Did you mean to just remove the CACHE PATH stuff since that's done all at the bottom?

Yeah, exactly! For the sake of deduplication.
But without `CACHE PATH` that line becomes
```
set(LIBR_HOME ${LIBR_HOME})
```
I don't think it makes any sense, just like https://github.com/microsoft/LightGBM/pull/2901#discussion_r393291279

---

ha oh true, good eye. That was unnecessary to include. Suggestion accepted!

---

just updated in https://github.com/microsoft/LightGBM/commit/9adc7a94fafeacf29e0bd88ffa320299406bbd65

I also rebased to `master` since a lot of R code was touched in #2803